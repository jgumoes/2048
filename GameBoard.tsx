import React, { useState } from "react"
import { View, useWindowDimensions, Text } from "react-native"
import { Gesture, GestureDetector, gestureHandlerRootHOC } from "react-native-gesture-handler"
import GameInfoBar from "./GameInfoBar"
import Grid, { Direction } from "./grid"
import { observer } from "mobx-react-lite"
import { appStyles, gameOverStyles, tileBackgroundColours, tileTextColours } from "./AppStyles"

type tileNumber_t = keyof typeof tileBackgroundColours;

function Tile({value, gridSides}:{value: number, gridSides: number}) {
  const styleIndex: tileNumber_t = Object.keys(tileBackgroundColours).includes(String(value))? String(value) as tileNumber_t : '4096'

  const valueString = value ? String(value) : ''

  const fontMultiplier = valueString.length <= 4 ? 20 : 15
  const fontSize = (gridSides * fontMultiplier)/245
  return(
    <View style={[appStyles.tile, tileBackgroundColours[styleIndex]]}>
      <Text style={[appStyles.tileText, tileTextColours[styleIndex], {fontSize: fontSize}]} >
        {valueString}
      </Text>
    </View>
  )
}

function GameOverOverlay({gridSides}: {gridSides: number}) {
  const fontSize = gridSides * 70 / 530
  return(
    <View style={gameOverStyles.overlay}>
      <Text style={[gameOverStyles.text, {fontSize: fontSize}]}>Game Over</Text>
    </View>
  )
}

/**
 * Creates and views a 4x4 grid
 * @param param0 
 * @returns 
 */
export const GridView4 = observer(({grid}: {grid: Grid}) => {
  
  const sidesX = useWindowDimensions().width * 0.95
  const sidesY = useWindowDimensions().height * 0.7
  const gridSides =  sidesX < sidesY ? sidesX : sidesY
  console.log("window sides: ", sidesX, sidesY)
  console.log("gridSides: ", gridSides)
  console.log("is game over: ", grid.isGameOver)

  return(
    <View style={[{height: gridSides, width: gridSides}]}>
      {grid.isGameOver && <GameOverOverlay gridSides={gridSides} />}
      <View style={[appStyles.gridView, {height: gridSides, width: gridSides}]}>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[0].map((value, index) => <Tile value={value} gridSides={gridSides} key={'0' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[1].map((value, index) => <Tile value={value} gridSides={gridSides} key={'1' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[2].map((value, index) => <Tile value={value} gridSides={gridSides} key={'2' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[3].map((value, index) => <Tile value={value} gridSides={gridSides} key={'3' + index} />)}
        </View>
      </View>
    </View>
  )
})

const WrappedGridView4 = gestureHandlerRootHOC(GridView4)

function findSwipeDirection({dx, dy}:{dx: number, dy:number}) {
  if(Math.abs(dx) > Math.abs(dy)){
    if(dx > 0){
      return Direction.right
    }
    return Direction.left
  }
  else{
    if(dy > 0){
      return Direction.down
    }
    return Direction.up
  }
}

export default function GameBoard({grid}: {grid: Grid}) {
  const [initialTouch, setInitialTouch] = useState({x:0, y:0})

  const swipeResponder = Gesture.Pan()
    .minDistance(100)
    .maxPointers(1)
    .onBegin((e) => {
      // when screen is first touched
      console.log("touch begin: ", Date.now())
      setInitialTouch({x: e.absoluteX, y: e.absoluteY})
    })
    .onStart((e) => {
      // when gesture is registered
      const direction = findSwipeDirection({
        dx: e.absoluteX - initialTouch.x, 
        dy: e.absoluteY - initialTouch.y
      })
      const now = Date.now()
      console.log("now: ", now)
      grid.swipe(direction) // this is nerfing performance
      console.log("swipe time: ", Date.now() - now)
      setInitialTouch({x: 0, y: 0})
    })

  console.log("rerendered: ", Date.now())
  return(
    <View>
      <GameInfoBar grid={grid} />
      <GestureDetector  gesture={swipeResponder}>
        <WrappedGridView4 grid={grid}/>
      </GestureDetector >
    </View>
  )
}