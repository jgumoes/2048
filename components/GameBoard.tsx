import { useEffect, useState } from "react"
import { View, useWindowDimensions, Text } from "react-native"
import { Gesture, GestureDetector, gestureHandlerRootHOC } from "react-native-gesture-handler"
import GameInfoBar from "./GameInfoBar"
import Grid, { Direction } from "../src/grid"
import { observer } from "mobx-react-lite"
import { appStyles, tileBackgroundColours, tileTextColours } from "../AppStyles"
import { GridOverlayState } from "../src/overlayState"
import { testIDs } from "../src/globalValues"
import GridOverlay from "./GridOverlay"

type tileNumber_t = keyof typeof tileBackgroundColours;

function Tile({value, gridSides}:{value: number, gridSides: number}) {
  const styleIndex: tileNumber_t = Object.keys(tileBackgroundColours).includes(String(value))? String(value) as tileNumber_t : '4096'

  const valueString = value ? String(value) : ''

  const fontMultiplier = valueString.length <= 4 ? 20 : 15
  const fontSize = (gridSides * fontMultiplier)/245
  return(
    <View style={[appStyles.tile, tileBackgroundColours[styleIndex]]}>
      <Text style={[appStyles.tileText, tileTextColours[styleIndex], {fontSize: fontSize}]} >
        {valueString}
      </Text>
    </View>
  )
}

export const gridSides = () => {
  const sidesX = useWindowDimensions().width * 0.95
  const sidesY = useWindowDimensions().height * 0.7
  return sidesX < sidesY ? sidesX : sidesY
}

/**
 * Creates and views a 4x4 grid
 * @param param0 
 */
export const NakedGridView4 = observer(({grid}: {grid: Grid}) => {
  const [primeGameWon, setPrimeGameWon] = useState(!grid.isGameWon)
  useEffect(()=>{
    console.log("GridView isGameWon useEffect triggered")
    if(primeGameWon && grid.isGameWon){
      console.log("GridView is showing gameWon overlay")
      setPrimeGameWon(false)
      GridOverlayState.showGameWonOverlay()
    }
  }, [grid.isGameWon])
  useEffect(()=>{
    console.log("GridView isGameOver useEffect triggered")
    if(grid.isGameOver){
      console.log("GridView is showing gameLost overlay")
      GridOverlayState.showGameLostOverlay()
    }
  }, [grid.isGameOver])
  
  return(
    <View style={[{height: gridSides(), width: gridSides()}]} testID={testIDs.gridView}>
      {!GridOverlayState.isHidden && <GridOverlay grid={grid} /> }
      <View style={[appStyles.gridView, {height: gridSides(), width: gridSides()}]}>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[0].map((value, index) => <Tile value={value} gridSides={gridSides()} key={'0' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[1].map((value, index) => <Tile value={value} gridSides={gridSides()} key={'1' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[2].map((value, index) => <Tile value={value} gridSides={gridSides()} key={'2' + index} />)}
        </View>
        <View style={appStyles.gridRow}>
          {grid.activeGrid[3].map((value, index) => <Tile value={value} gridSides={gridSides()} key={'3' + index} />)}
        </View>
      </View>
    </View>
  )
})

export const GridView4 = gestureHandlerRootHOC(NakedGridView4)

function findSwipeDirection({dx, dy}:{dx: number, dy:number}) {
  if(Math.abs(dx) > Math.abs(dy)){
    if(dx > 0){
      return Direction.right
    }
    return Direction.left
  }
  else{
    if(dy > 0){
      return Direction.down
    }
    return Direction.up
  }
}

const GameBoard = observer(({grid}: {grid: Grid}) => {
  const [initialTouch, setInitialTouch] = useState({x:0, y:0})

  const swipeResponder = Gesture.Pan()
  // TODO: mock this class to create swipe function
  // then create a function that triggers the swipe 
    .minDistance(100)
    .maxPointers(1)
    .onBegin((e) => {
      // when screen is first touched
      console.log("touch begin: ", Date.now())
      setInitialTouch({x: e.absoluteX, y: e.absoluteY})
    })
    .onStart((e) => {
      // when gesture is registered
      const direction = findSwipeDirection({
        dx: e.absoluteX - initialTouch.x, 
        dy: e.absoluteY - initialTouch.y
      })
      const now = Date.now()
      console.log("now: ", now)
      if(GridOverlayState.isHidden){grid.swipe(direction)} // TODO: this is nerfing performance
      console.log("swipe time: ", Date.now() - now)
      setInitialTouch({x: 0, y: 0})
    })
  
  console.log("rerendered: ", Date.now())
  return(
    <View>
      <GameInfoBar grid={grid} />
      <GestureDetector  gesture={swipeResponder}>
        <GridView4 grid={grid} />
      </GestureDetector >
    </View>
  )
})

export default GameBoard