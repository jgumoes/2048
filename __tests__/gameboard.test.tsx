import { render, screen } from '@testing-library/react-native'

import Grid from '../grid';

import * as importedTestGrids from './testGrids.json'
import GameBoard, { GridView4 } from '../GameBoard';

jest.disableAutomock()

describe('GridView4', () => {
  test('renders correctly', () => {
    const testGrid = new Grid({grid: importedTestGrids.resultGrids.down[0]});
    render(<GridView4 grid={testGrid} />);
    expect(screen.toJSON()).toMatchSnapshot()
  })
})