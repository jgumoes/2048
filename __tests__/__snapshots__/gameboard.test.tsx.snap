// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridView4 renders correctly 1`] = `
<View
  style={
    [
      {
        "height": 712.5,
        "width": 712.5,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "backgroundColor": "rgb(187, 173, 160)",
          "borderRadius": 10,
          "justifyContent": "space-evenly",
        },
        {
          "height": 712.5,
          "width": 712.5,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(238, 228, 218)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 149, 99)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          16
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(246, 93, 59)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          64
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 124, 95)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          32
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 124, 95)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          32
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "height": 712.5,
        "width": 712.5,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "backgroundColor": "rgb(187, 173, 160)",
          "borderRadius": 10,
          "justifyContent": "space-evenly",
        },
        {
          "height": 712.5,
          "width": 712.5,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(238, 228, 218)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(214, 205, 196)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(236, 224, 200)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(119, 110, 101)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 149, 99)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          16
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "height": "22%",
          "justifyContent": "space-evenly",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(246, 93, 59)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          64
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(242, 177, 121)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          8
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 124, 95)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          32
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "justifyContent": "center",
              "textAlign": "center",
              "width": "22%",
            },
            {
              "backgroundColor": "rgb(245, 124, 95)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "color": "rgb(249, 246, 241)",
              },
              {
                "fontSize": 58.16326530612245,
              },
            ]
          }
        >
          32
        </Text>
      </View>
    </View>
  </View>
</View>
`;
